substrate/frame/staking/src/pallet/mod.rs

```diff
@@ -1980,6 +1983,115 @@ pub mod pallet {
 
 			Ok(Some(T::WeightInfo::deprecate_controller_batch(controllers.len() as u32)).into())
 		}
+
+		/// Restores the state of a ledger which is in an inconsistent state.
+		///
+		/// The requirements to restore a ledger are the following:
+		/// * The stash is bonded;
+		/// * If the stash has an associated ledger, its state must be inconsistent;
+		/// * If the ledger is not corrupted *but* its staking lock is out of sync.
+		///
+		/// The `maybe_*` input parameters will overwrite the corresponding data and metadata of the
+		/// ledger associated with the stash. If the input parameters are not set, the ledger will
+		/// be reset with the current values.
+		///
+		/// Upon successful execution, this extrinsic will *recreate* the ledger based on its past
+		/// state and/or the `maybe_controller` and `maybe_total` input parameters.
+		#[pallet::call_index(29)]
+		#[pallet::weight(T::WeightInfo::restore_ledger())]
+		pub fn restore_ledger(
```

Comment:
So this is like a throw away function right once the issues are fixed?

Reply:
I think you earlier tried the approach where staking admin provides some values for the ledger, and this function applies them if it could verify its integrity. Would that approach not work? Seems a simpler and more general approach and potentially useful for future as well.

Reply:
Also would avoid bug if we simply this fn (seems too many code branches). OTOH it is gated by StakingAdmin, so safe-ish.

Reply:
> Would that approach not work? Seems a simpler and more general approach and potentially useful for future as well.

It would, but since that's more loosely defined and just overwrites the ledger/bond/lock values, it's less safe (even when behind staking admin). So I see the options to reset the ledgers as one of (or a spectrum in-between of):

1. Have a `reset_ledger` or `set_ledger` call that basically re-writes any value in storage related to staking. This is a glorified `set_storage` for (some) staking storage items.
2. Have an extrinsic that will restore the stash/ledger state relying as much as possible on the on-chain state to learn what's the corruption case and its fix and to use as much of the current on-chain as possible (e.g. using the current locks of a stash to reset its ledger).

After mapping all the potential corruption cases, the second option seemed better and safer to me. I believe that it would be also easier to make the case for governance to review/accept the proposal (it relies more on current on-chain data, we won't need to present proofs for every single value re-written per ledger).

The option 2. loses in generality. But I think that's OK and perhaps we should rely on `set_code` or implement a `force_unbond/clean_ledger` (which does not rebond, only cleans) in the future.